pkg/scheduler/reconcile_on_synctarget.go

```diff
@@ -19,20 +19,254 @@ package scheduler
 import (
 	"context"
 
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
 	"k8s.io/klog/v2"
 
 	kcpcache "github.com/kcp-dev/apimachinery/v2/pkg/cache"
+	schedulingv1alpha1 "github.com/kcp-dev/kcp/pkg/apis/scheduling/v1alpha1"
+	workloadv1alpha1 "github.com/kcp-dev/kcp/pkg/apis/workload/v1alpha1"
+	"github.com/kcp-dev/logicalcluster/v3"
+
+	edgev1alpha1 "github.com/kcp-dev/edge-mc/pkg/apis/edge/v1alpha1"
 )
 
-func (c *controller) reconcileOnSyncTarget(ctx context.Context, key string) error {
+func (c *controller) reconcileOnSyncTarget(ctx context.Context, stKey string) error {
 	logger := klog.FromContext(ctx)
-	ws, _, name, err := kcpcache.SplitMetaClusterNamespaceKey(key)
+	stws, _, stName, err := kcpcache.SplitMetaClusterNamespaceKey(stKey)
 	if err != nil {
-		logger.Error(err, "invalid key")
+		logger.Error(err, "invalid SyncTarget key")
 		return err
 	}
-	logger = logger.WithValues("workspace", ws, "syncTarget", name)
+	logger = logger.WithValues("syncTargetWorkspace", stws, "syncTarget", stName)
 	logger.V(2).Info("reconciling")
 
+	/*
+		On synctarget change:
+		1a) find all ep(s) that used st
+
+		2a) find all loc(s) that selecting st, and update store
+		2b) find all ep(s) that using st
+
+		3a) for each of its obsolete ep, remove all sp(s) that has st
+		3b) for each of its ongoing ep, update sp(s)
+		3c) for each of its new ep, add sp(s)
+
+		Need data structure:
+		- map from a synctarget to its locations, say 'locsBySelectedSt'
+	*/
+
+	st, err := c.synctargetLister.Cluster(stws).Get(stName)
+	if err != nil {
+		logger.Error(err, "syncTarget not found in local cache")
+		return nil
+	}
+
+	store.l.Lock()
+	defer store.l.Unlock() // TODO(waltforme): Is it safe to shorten the critical section?
+
+	// 1a)
+	epsUsedSt := store.findEpsUsedSt(stKey)
+
+	// 2a)
+	locsInStws, err := c.locationLister.Cluster(stws).List(labels.Everything())
+	if err != nil {
+		logger.Error(err, "failed to list Locations")
+		return err
+	}
+	locsSelectingSt, err := filterLocsBySt(locsInStws, st)
+	if err != nil {
+		logger.Error(err, "failed to find Locations for SyncTarget")
+		return err
+	}
+
+	store.locsBySelectedSt[stKey] = packLocKeys(locsSelectingSt)
+
+	// 2b)
+	epsUsingSt := map[string]empty{}
+	for _, loc := range locsSelectingSt {
+		locKey, _ := kcpcache.MetaClusterNamespaceKeyFunc(loc)
+		eps := store.epsBySelectedLoc[locKey]
+		epsUsingSt = unionTwo(epsUsingSt, eps)
+	}
```

Comment:
This is map/reduce, for which you could write one generic definition rather than spell it out each time.